ncm-network/src/main/perl/nmstate.pm

```diff
@@ -0,0 +1,720 @@
+Readonly my $NM_RESOLV_YML => "/etc/nmstate/resolv.yml";
+Readonly my $NM_DROPIN_CFG_FILE => "/etc/NetworkManager/conf.d/90-quattor.conf";
+
+# generate the correct fake yaml boolean value so TextRender can convert it in a yaml boolean
+Readonly my $YTRUE => $EDG::WP4::CCM::TextRender::ELEMENT_CONVERT{yaml_boolean}->(1);
+Readonly my $YFALSE => $EDG::WP4::CCM::TextRender::ELEMENT_CONVERT{yaml_boolean}->(0);
+
+use constant IFCFG_DIR => "/etc/nmstate";
+
+sub iface_filename
+{
+    my ($self, $iface) = @_;
+    return $self->IFCFG_DIR . "/$iface.yml";
+}
+
+# Given the configuration in nmconnection
+# Determine if this is a valid interface for ncm-network to manage,
+# Return arrayref tuple [interface name, ifdown/ifup name] when valid,
+# undef otherwise.
+sub is_valid_interface
+{
+    my ($self, $filename) = @_;
+
+    # Very primitive, based on regex only
+    # matchs eth0.yml bond0.yml, or bond0.101.yml
+    if (
+        $filename =~ m{
+            # Filename is either right at the beginning or following a slash
+            (?: \A | / )
+            # $1 will capture for example:
+            # eth0  bond1  eth0.101  bond0.102
+            ( \w+ \d+ (?: \. \d+ )? )
+            # Suffix (not captured)
+            \. yml \z
+        }x
+    ) {
+        # name and id for nmstate, this will make connection id and name the same.
+        my $name = $1;
+        # previous is_valid_interface reuqires suffix, not concerned about this in nmstate so just return $name as suffix too.
+        return [$name, $name];
+    } else {
+        return;
+    };
+}
+
+# By default, NetworkManager on Red Hat Enterprise Linux (RHEL) 8+ dynamically updates the /etc/resolv.conf
+# file with the DNS settings from active NetworkManager connection profiles. we manage this using ncm-resolver.
+# so disable this unless nm_manage_dns = true. resolver details can be set using nmstate but not doing this now.
+sub disable_nm_manage_dns
+{
+    my ($self, $manage_dns, $nwsrv) = @_;
+    my @data = ('[main]');
+    
+    if ( $manage_dns ) {
+        # set nothing, will use default.
+        $self->verbose("Networkmanager defaults will be used to manage resolv.conf");
+    } else {
+        push @data, 'dns=none';
+        $self->verbose("Configuring networkmanager not to manage resolv.conf");
+    }
+    my $fh = CAF::FileWriter->new($NM_DROPIN_CFG_FILE, mode => oct(444), log => $self, keeps_state => 1);
+    print $fh join("\n", @data, '');
+    if ($fh->close()) {
+        $self->info("File $NM_DROPIN_CFG_FILE changed, reload network");
+        $nwsrv->reload();
+    };
+}
+
+# return hasref of ipv4 policy rule
+sub make_nm_ip_rule
+{
+    my ($self, $device, $rules, $routing_table_hash) = @_;
+
+    my @rule_entry;
+    foreach my $rule (@$rules) {
+        my %thisrule;
+        my $priority = 100;
+        $priority = $rule->{priority} if $rule->{priority};
+        $thisrule{'family'} = "ipv4";
```

Comment:
quotes are not needed in the key in perl, unless eg when using a `-`, then perl things this is a subtraction. so remove it here